File: CrackingTheCodingInterview/ch04-TreesAndGraphs/firstCommonAncestor.js
/*
  design an algorithm to find the first common ancestor of two
  nodes in a binary tree
  avoid storing additional nodes in a data structure
*/

function firstCommonAncestor(root, a, b) {
  // error check - one node is not in tree
  if (!isDescendant(root, a) || !isDescendant(root, b)) return null;

  return ancestorHelper(root, a, b);
}

function ancestorHelper(node, a, b) {
  if (node === a || node === b) return node;

  const aOnLeft = isDescendant(node, a);
  const bOnLeft = isDescendant(node, b);

  if (aOnLeft !== bOnLeft) { // a and b on different sides
    return node;
  } else if (aOnLeft) { // a and b on left side
    return ancestorHelper(node.left, a, b)
  } else { // a and b on right side
    return ancestorHelper(node.right, a, b)
  }
}

function isDescendant(ancestor, child) {
  if (ancestor === child) return true;
  if (!ancestor) return false;
  return isDescendant(ancestor.left, child) ||
    isDescendant(ancestor.right, child);
}

File: CrackingTheCodingInterview/ch04-TreesAndGraphs/firstCommonAncestor.py
# Design an algorithm to find the first common ancestor of two
# nodes in a binary tree
# Avoid storing additional nodes in a data structure


def first_common_ancestor(root, a, b):
    # if a or b not in tree
    if not is_descendant(root, a) or not is_descendant(root, b):
        return None

    def helper(node, a, b):
        if node == a or node == b:
            return node

        a_on_left = is_descendant(node.left, a)
        b_on_left = is_descendant(node.left, b)

        # if on different sides
        if a_on_left != b_on_left:
            return node
        elif a_on_left:
            return helper(node.left, a, b)
        else:
            return helper(node.right, a, b)

    return helper(root, a, b)


def is_descendant(ancestor, child):
    if ancestor == child:
        return True
    if not ancestor:
        return False

    return is_descendant(ancestor.left, child) or \
        is_descendant(ancestor.right, child)